src/Symfony/Bundle/SecurityBundle/Resources/config/collectors.xml

```diff
@@ -14,6 +14,7 @@
             <argument type="service" id="security.logout_url_generator" />
             <argument type="service" id="security.access.decision_manager" />
             <argument type="service" id="security.firewall.map" />
+            <argument type="service" id="security.firewall" />
```

Comment:
If you don't inject explicitly `debug.security.firewall` here, you can't typehint `TraceableFirewallListener`, but should add an `instanceof` check in the collector, otherwise it'll break in non debug mode when the `TraceableFirewallListener` does not decorates `security.firewall` (I still don't get why we would have collectors in non-debug mode though...).

IMHO, it's better to keep your collector as you did, but inject `debug.security.firewall` with `on-invalid="null"` here.